### PR TITLE
New logging capability and some other updates

### PR DESCRIPTION
**Pull request by Ian-Clowes**
Mainly focussed on adding logging that can be usable by average users and developers alike.

Also includes some other updates to memory tracing, a few bugs that were spotted while getting things working, etc.

### CODE REVIEW

**Comment by PolynomialDivision** [comment]
Wow awesome thanks! Your commit history is a bit broken. I can fix it for you if you like?

**Comment by Ian-Clowes** [comment]
> Wow awesome thanks! Your commit history is a bit broken. I can fix it for you if you like?

I think my commit history is always broken!  I'm surprised how much I still struggle with git - although it can be friendly to me sometimes.

Yes, I'd be delighted to have things fixed.

**Comment by PolynomialDivision** [comment]
Alright. :) I will do this and replace the PR. I guess you would benefit before doing any further changes, something that you save your work and reset your git history to the master.

Probably something like

    git remote add upstream https://github.com/berlin-open-wireless-lab/DAWN.git
    
then

    git fetch upstream

and then

    git reset --hard upstream/master
    
Then you are equal again with master branch. **BUT BE CAREFUL THAT WILL DELETE EVERYTHING YOU HAVE IN YOUR GIT! SAVE EVERYTHING BEFORE** :)

**Comment by Ian-Clowes** [comment]
> Alright. :) I will do this and replace the PR. I guess you would benefit before doing any further changes, something that you save your work and reset your git history to the master.
> 
> Probably something like
> ...
> Then you are equal again with master branch. **BUT BE CAREFUL THAT WILL DELETE EVERYTHING YOU HAVE IN YOUR GIT! SAVE EVERYTHING BEFORE** :)

Can I do that for the github fork via web clicks?  I can then discard all my local work areas and start afresh from that fork.

**Comment by PolynomialDivision** [comment]
It would be easier to clone and do that locally. You can also delete your old fork, and fork again. ;) But please don't do that, since I currtenly working with the diff to your master branch. However, it is harder than I thought do seperate that into single commits. The treewide logging change, produces a huge diff. So there are all these little changes you made, like deleting `decide_function`, setting pointers to NULL, ..., that are hard to detect. I started here, but I need some time, to go through it (it is not finished):
https://github.com/berlin-open-wireless-lab/DAWN/commits/new-pr

I tried to grep for only the dawnlog diffs, but I was not successfull in creating with that a patch file. However I think that is the easiest method, to create a patchfile from the diff that only contains the logging changes, and afterwards apply all the other changes. If anyone has an idea, how to do this in an elegant way, I am happy to hear some suggestions. :)

**Comment by Ian-Clowes** [comment]
> I tried to grep for only the dawnlog diffs, but I was not successfull in creating with that a patch file. However I think that is the easiest method, to create a patchfile from the diff that only contains the logging changes, and afterwards apply all the other changes. If anyone has an idea, how to do this in an elegant way, I am happy to hear some suggestions. :)

I thought about doing the same myself.  I'm more familar with the changes of course, so if you want to nominate a baseline I can create 3 or more patches that apply in sequence to create the overall PR, such as one each for "dawnlog()", "memory auditing" and "other".

**Comment by PolynomialDivision** [comment]
That would be awesome! I would prefer first adding the logging, since this is not so error prone and very important. The other changes on top. I know it looks a bit nitpicking but to have a good commit history, with strong ocmmit messages, helps everyone else to understand the changes and really helps in case of bugfixing and bisecting.

**Comment by Ian-Clowes** [comment]
> That would be awesome! I would prefer first adding the logging, since this is not so error prone and very important. The other changes on top. I know it looks a bit nitpicking but to have a good commit history, with strong ocmmit messages, helps everyone else to understand the changes and really helps in case of bugfixing and bisecting.

OK, what shall I take as a basline?  Current HEAD on master?  May sound like a daft question - but I'd like to get it right :)

**Comment by PolynomialDivision** [comment]
With baseline, you mean the commits you set on top? Yeah, please use the master branch of DAWN origin repository. :)
I saved now your master branch here:
https://github.com/berlin-open-wireless-lab/DAWN/tree/save-ian

Now you can do whatever you want, and we won't lose something (at least I hope it, except you have changes in some branch).

**Comment by Ian-Clowes** [comment]
Here are a set of patches that should apply cleanly with `patch -p1` in the order below.

LOGGING: Add dawnlog()
Add dawnlog_* functions and macros to convert printf() family and perror() logging to syslog() family
Removed unnecessary sprintf() for building log strings (embed format directly)
Add local config settings for log level
Add command line parameters for log level and destination
Set default log level to suppress a lot of previously noisy messages
Restore some previously removed noisy messages as DEBUG level in case they help in future
Eliminate DAWN_NO_OUTPUT static code checks which are no longer used

MEM: TIghten up some memory handling to help spot errors
Set pointers to NULL after free() to help force out memory handling errors
Add some extra memory / resource tracking to try and chase out latent bugs / leaks
Fixed a couple of memory traces that were misreporting

MAINT: General revisions to maintain code (no functional changes intended)
Removed unnecessary linked-list length checks
Fixed some typos on function names / comments
Changed how test_storage forces SEGV due to new compiler warnings

FIX1: Bug that could proably cause ineffciency or confusion in multi-SSID environments
Adjusted erroneus match test in ap_array_insert()

FIX2: Bug that was preventing kicking working due to mishandling of bandwidth discovery
Fixed bug in use of get_bandwidth_iwinfo() in AP kicking
Fix rounding of transmission rate calculations in get_bandwidth_iwinfo()
Restructure of get_bandwidth...() while finding bug

FEAT1: Refresh of kicking algorithm to minimise messages and streamline some code
Removed decide_function() which made some logic hard to follow
Removed kick_client() function to streamline code
Revised kicking algorithm code to give better user messages

[DAWN-logging-f.txt](https://github.com/berlin-open-wireless-lab/DAWN/files/7818621/DAWN-logging-f.txt)
[DAWN-mem-f.txt](https://github.com/berlin-open-wireless-lab/DAWN/files/7818622/DAWN-mem-f.txt)
[DAWN-maint-f.txt](https://github.com/berlin-open-wireless-lab/DAWN/files/7818623/DAWN-maint-f.txt)
[DAWN-fix1-f.txt](https://github.com/berlin-open-wireless-lab/DAWN/files/7818625/DAWN-fix1-f.txt)
[DAWN-fix2-f.txt](https://github.com/berlin-open-wireless-lab/DAWN/files/7818624/DAWN-fix2-f.txt)
[DAWN-feat1-f.txt](https://github.com/berlin-open-wireless-lab/DAWN/files/7818626/DAWN-feat1-f.txt)

**Comment by PolynomialDivision** [comment]
Replaced by:
https://github.com/berlin-open-wireless-lab/DAWN/pull/157

Thanks a lot!

**Comment by marazmarci** [comment]
I think you meant "Replaced by: #158", since #157 is this pull request that you just closed 😄

**Comment by PolynomialDivision** [comment]
Yes. Thanks. ;)
